## Session 4 : Feature engineering - Home Credit Risk

##### Student: Katayoun B.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline    
import seaborn as sns
import glob

###  Understanding the tables and loading all

In [2]:
def load_data(path):
    data_path = path 
    df_files = glob.glob(data_path+"*.csv")
    df_files = sorted(df_files, key=str.lower)
    return df_files

In [3]:
df_files = load_data('/Users/katy/desktop/ml/homework3/home_credit_risk/')
df_files

['/Users/katy/desktop/ml/homework3/home_credit_risk/application_train.csv',
 '/Users/katy/desktop/ml/homework3/home_credit_risk/bureau.csv',
 '/Users/katy/desktop/ml/homework3/home_credit_risk/bureau_balance.csv',
 '/Users/katy/desktop/ml/homework3/home_credit_risk/credit_card_balance.csv',
 '/Users/katy/desktop/ml/homework3/home_credit_risk/installments_payments.csv',
 '/Users/katy/desktop/ml/homework3/home_credit_risk/POS_CASH_balance.csv',
 '/Users/katy/desktop/ml/homework3/home_credit_risk/previous_application.csv']

In [4]:
csvs = len(df_files)
csvs

7

In [5]:
# creating datafram for each csv
for x in range(csvs):
    if x == 0:
        print(df_files[0])
        main_df = pd.read_csv(df_files[0])
        #print(train_df)
    if x == 1:
        print(df_files[1])
        bureau_df = pd.read_csv(df_files[1])
    if x == 2:
        print(df_files[2])
        bureau_balance_df = pd.read_csv(df_files[2])
    if x == 3:
        print(df_files[3])
        credit_balance_df = pd.read_csv(df_files[3])
    if x == 4:
        print(df_files[4])
        installments_df = pd.read_csv(df_files[4])
    if x == 5:
        print(df_files[5])
        pos_cash_df = pd.read_csv(df_files[5])
    if x == 6:
        print(df_files[6])
        prev_df = pd.read_csv(df_files[6])

/Users/katy/desktop/ml/homework3/home_credit_risk/application_train.csv
/Users/katy/desktop/ml/homework3/home_credit_risk/bureau.csv
/Users/katy/desktop/ml/homework3/home_credit_risk/bureau_balance.csv
/Users/katy/desktop/ml/homework3/home_credit_risk/credit_card_balance.csv
/Users/katy/desktop/ml/homework3/home_credit_risk/installments_payments.csv
/Users/katy/desktop/ml/homework3/home_credit_risk/POS_CASH_balance.csv
/Users/katy/desktop/ml/homework3/home_credit_risk/previous_application.csv


In [6]:
print(main_df.shape)
main_df.head()

(307511, 122)


SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 122 columns]

In [7]:
#print(main_df.columns.values)

### Assignment 1: complete feature analysis for main table
        

1. To create a new feature between 'AMT_CREDIT', 'AMT_INCOME_TOTAL'
2. I want to call this new feature 'HIGH_INCOME', One(1) for 'HIGH_INCOME' and Zero(0) for less than mean value wich is 3.9

In [8]:
main_df[['AMT_CREDIT','AMT_INCOME_TOTAL']]

AMT_CREDIT  AMT_INCOME_TOTAL
0         406597.5          202500.0
1        1293502.5          270000.0
2         135000.0           67500.0
3         312682.5          135000.0
4         513000.0          121500.0
...            ...               ...
307506    254700.0          157500.0
307507    269550.0           72000.0
307508    677664.0          153000.0
307509    370107.0          171000.0
307510    675000.0          157500.0

[307511 rows x 2 columns]

In [9]:
main_df['HIGH_INCOME'] = main_df['AMT_CREDIT'] / main_df['AMT_INCOME_TOTAL']
main_df['HIGH_INCOME'].mean()

3.9575702380626114

In [10]:
main_df['HIGH_INCOME'] = main_df['HIGH_INCOME'].apply(lambda x : 0 if x < 3 else 1)

In [11]:
main_df[['SK_ID_CURR', 'AMT_CREDIT', 'AMT_INCOME_TOTAL', 'HIGH_INCOME']]

SK_ID_CURR  AMT_CREDIT  AMT_INCOME_TOTAL  HIGH_INCOME
0           100002    406597.5          202500.0            0
1           100003   1293502.5          270000.0            1
2           100004    135000.0           67500.0            0
3           100006    312682.5          135000.0            0
4           100007    513000.0          121500.0            1
...            ...         ...               ...          ...
307506      456251    254700.0          157500.0            0
307507      456252    269550.0           72000.0            1
307508      456253    677664.0          153000.0            1
307509      456254    370107.0          171000.0            0
307510      456255    675000.0          157500.0            1

[307511 rows x 4 columns]

1. I create a new feature called 'RISKY_NEW_JOB_FLAG', I assume if they switched the phone number recently that means they changed the job recently.
2. if this number is bigger 180 days ( 6 months - max number of months to pass probation ), ignore this otherwise flag as a risky with new job 'RISKY_NEW_JOB_FLAG'.
3. one(1) for risky, Zero(0) for not-risky.

In [12]:
main_df['DAYS_LAST_PHONE_CHANGE']

0        -1134.0
1         -828.0
2         -815.0
3         -617.0
4        -1106.0
           ...  
307506    -273.0
307507       0.0
307508   -1909.0
307509    -322.0
307510    -787.0
Name: DAYS_LAST_PHONE_CHANGE, Length: 307511, dtype: float64

In [13]:
main_df['RISKY_NEW_JOB_FLAG'] = main_df['DAYS_LAST_PHONE_CHANGE'].abs()
main_df['RISKY_NEW_JOB_FLAG'] = main_df['RISKY_NEW_JOB_FLAG'].apply(lambda x : 0 if x < 180 else 1)

In [14]:
main_df[['SK_ID_CURR', 'RISKY_NEW_JOB_FLAG']]
# Then we might drop that col, just an idea

SK_ID_CURR  RISKY_NEW_JOB_FLAG
0           100002                   1
1           100003                   1
2           100004                   1
3           100006                   1
4           100007                   1
...            ...                 ...
307506      456251                   1
307507      456252                   0
307508      456253                   1
307509      456254                   1
307510      456255                   1

[307511 rows x 2 columns]

Create a new feature called 'IF_EMPLOYED', zero (0) is not-employed any more, one (1) for employed. 

In [15]:
main_df['IF_EMPLOYED'] = main_df['DAYS_EMPLOYED'].apply(lambda x : 0 if x > 0 else 1)
main_df[['SK_ID_CURR', 'IF_EMPLOYED']]

SK_ID_CURR  IF_EMPLOYED
0           100002            1
1           100003            1
2           100004            1
3           100006            1
4           100007            1
...            ...          ...
307506      456251            1
307507      456252            0
307508      456253            1
307509      456254            1
307510      456255            1

[307511 rows x 2 columns]

In [16]:
# check if flaged them correctly
main_df.loc[main_df.IF_EMPLOYED == 0]

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
8           100011       0         Cash loans           F            N   
11          100015       0         Cash loans           F            N   
23          100027       0         Cash loans           F            N   
38          100045       0         Cash loans           F            N   
43          100050       0         Cash loans           F            N   
...            ...     ...                ...         ...          ...   
307469      456209       0         Cash loans           F            N   
307483      456227       0         Cash loans           F            N   
307487      456231       0         Cash loans           M            N   
307505      456249       0         Cash loans           F            N   
307507      456252       0         Cash loans           F            N   

       FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  \
8                    Y             0        112500.000   1019610.0   
11                   Y             0         38419.155    148365.0   
23                   Y             0         83250.000    239850.0   
38                   Y             0         99000.000    247275.0   
43                   Y             0        108000.000    746280.0   
...                ...           ...               ...         ...   
307469               Y             0        202500.000    703728.0   
307483               Y             0         99000.000    247275.0   
307487               Y             0        117000.000   1071909.0   
307505               Y             0        112500.000    225000.0   
307507               Y             0         72000.000    269550.0   

        AMT_ANNUITY  ...  FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR  \
8           33826.5  ...                 0                        0.0   
11          10678.5  ...                 0                        0.0   
23          23850.0  ...                 0                        0.0   
38          17338.5  ...                 0                        0.0   
43          42970.5  ...                 0                        0.0   
...             ...  ...               ...                        ...   
307469      29943.0  ...                 0                        0.0   
307483      16479.0  ...                 0                        0.0   
307487      31473.0  ...                 0                        0.0   
307505      22050.0  ...                 0                        0.0   
307507      12001.5  ...                 0                        NaN   

       AMT_REQ_CREDIT_BUREAU_DAY AMT_REQ_CREDIT_BUREAU_WEEK  \
8                            0.0                        0.0   
11                           0.0                        0.0   
23                           0.0                        0.0   
38                           0.0                        0.0   
43                           0.0                        0.0   
...                          ...                        ...   
307469                       0.0                        0.0   
307483                       0.0                        0.0   
307487                       0.0                        0.0   
307505                       0.0                        0.0   
307507                       NaN                        NaN   

       AMT_REQ_CREDIT_BUREAU_MON AMT_REQ_CREDIT_BUREAU_QRT  \
8                            0.0                       0.0   
11                           0.0                       0.0   
23                           0.0                       0.0   
38                           0.0                       0.0   
43                           0.0                       0.0   
...                          ...                       ...   
307469                       1.0                       0.0   
307483                       0.0                       0.0   
307487                       0.0                       1.0   
307505                  

1. Create a new feature called 'DAYS_EMPLOYED_PCT' to show the number of work experience (senior or junior). 
2. calculating the percentage of 'DAYS_EMPLOYED' and 'DAYS_BIRTH', 
3. if this nubmer is bigger than > 0.2 I assume that persion is a senior or at least has enough work experience. 

In [17]:
main_df['DAYS_EMPLOYED_PCT'] = main_df['DAYS_EMPLOYED'] / main_df['DAYS_BIRTH']

In [18]:
main_df['DAYS_EMPLOYED_PCT'] = main_df['DAYS_EMPLOYED_PCT'].abs()
main_df['DAYS_EMPLOYED_PCT'] = round(main_df['DAYS_EMPLOYED_PCT'])
main_df[['SK_ID_CURR', 'DAYS_EMPLOYED_PCT']]

SK_ID_CURR  DAYS_EMPLOYED_PCT
0           100002                0.0
1           100003                0.0
2           100004                0.0
3           100006                0.0
4           100007                0.0
...            ...                ...
307506      456251                0.0
307507      456252               18.0
307508      456253                1.0
307509      456254                0.0
307510      456255                0.0

[307511 rows x 2 columns]

In [19]:
main_df['DAYS_EMPLOYED_PCT'] = main_df['DAYS_EMPLOYED_PCT'].apply(lambda x : 0 if x < 0.2 else 1)
main_df[['SK_ID_CURR', 'DAYS_EMPLOYED_PCT']]

SK_ID_CURR  DAYS_EMPLOYED_PCT
0           100002                  0
1           100003                  0
2           100004                  0
3           100006                  0
4           100007                  0
...            ...                ...
307506      456251                  0
307507      456252                  1
307508      456253                  1
307509      456254                  0
307510      456255                  0

[307511 rows x 2 columns]

### Assignment 2: team up to expand more features with crazy brain storming

# bureau_df

In [20]:
print(bureau_df.shape)
bureau_df.head()

(1716428, 17)


SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  DAYS_CREDIT  \
0      215354       5714462        Closed      currency 1         -497   
1      215354       5714463        Active      currency 1         -208   
2      215354       5714464        Active      currency 1         -203   
3      215354       5714465        Active      currency 1         -203   
4      215354       5714466        Active      currency 1         -629   

   CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  DAYS_ENDDATE_FACT  \
0                   0               -153.0             -153.0   
1                   0               1075.0                NaN   
2                   0                528.0                NaN   
3                   0                  NaN                NaN   
4                   0               1197.0                NaN   

   AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  AMT_CREDIT_SUM  \
0                     NaN                   0         91323.0   
1                     NaN                   0        225000.0   
2                     NaN                   0        464323.5   
3                     NaN                   0         90000.0   
4                 77674.5                   0       2700000.0   

   AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  AMT_CREDIT_SUM_OVERDUE  \
0                  0.0                   NaN                     0.0   
1             171342.0                   NaN                     0.0   
2                  NaN                   NaN                     0.0   
3                  NaN                   NaN                     0.0   
4                  NaN                   NaN                     0.0   

       CREDIT_TYPE  DAYS_CREDIT_UPDATE  AMT_ANNUITY  
0  Consumer credit                -131          NaN  
1      Credit card                 -20          NaN  
2  Consumer credit                 -16          NaN  
3      Credit card                 -16          NaN  
4  Consumer credit                 -21          NaN

In [21]:
bureau_df.loc[bureau_df.CREDIT_ACTIVE=='Closed']

SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  DAYS_CREDIT  \
0            215354       5714462        Closed      currency 1         -497   
7            162297       5714469        Closed      currency 1        -1896   
8            162297       5714470        Closed      currency 1        -1146   
11           162297       5714473        Closed      currency 1        -2456   
14           238881       5714482        Closed      currency 1         -318   
...             ...           ...           ...             ...          ...   
1716422      375755       5057742        Closed      currency 1        -2648   
1716424      100044       5057754        Closed      currency 1        -2648   
1716425      100044       5057762        Closed      currency 1        -1809   
1716426      246829       5057770        Closed      currency 1        -1878   
1716427      246829       5057778        Closed      currency 1         -463   

         CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  DAYS_ENDDATE_FACT  \
0                         0               -153.0             -153.0   
7                         0              -1684.0            -1710.0   
8                         0               -811.0             -840.0   
11                        0               -629.0             -825.0   
14                        0               -187.0             -187.0   
...                     ...                  ...                ...   
1716422                   0              31129.0             -189.0   
1716424                   0              -2433.0            -2493.0   
1716425                   0              -1628.0             -970.0   
1716426                   0              -1513.0            -1513.0   
1716427                   0                  NaN             -387.0   

         AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  AMT_CREDIT_SUM  \
0                           NaN                   0        91323.00   
7                       14985.0                   0        76878.45   
8                           0.0                   0       103007.70   
11                          NaN                   0       675000.00   
14                          NaN                   0            0.00   
...                         ...                 ...             ...   
1716422                     NaN                   0       202500.00   
1716424                  5476.5                   0        38130.84   
1716425                     NaN                   0        15570.00   
1716426                     NaN                   0        36000.00   
1716427                     NaN                   0        22500.00   

         AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  AMT_CREDIT_SUM_OVERDUE  \
0                        0.0                   NaN                     0.0   
7                        0.0                   0.0                     0.0   
8                        0.0                   0.0                     0.0   
11                       0.0                   0.0                     0.0   
14                       0.0                   0.0                     0.0   
...                      ...                   ...                     ...   
1716422                  0.0                   NaN                     0.0   
1716424                  0.0                   0.0                     0.0   
1716425                  NaN                   NaN                     0.0   
1716426                  0.0                   0.0                     0.0   
1716427                  0.0                   NaN                     0.0   

             CREDIT_TYPE  DAYS_CREDIT_UPDATE  AMT_ANNUITY  
0        Consumer credit                -131          NaN  
7        Consumer credit               -1710          NaN  
8        Consumer credit                -840          NaN  
11       Consumer credit                -706          NaN  
14           Credit card                -185          NaN  
...                  ...                 ... 

In [22]:
bureau_df.groupby('SK_ID_CURR')['SK_ID_BUREAU'].size()

SK_ID_CURR
100001     7
100002     8
100003     4
100004     2
100005     3
          ..
456249    13
456250     3
456253     4
456254     1
456255    11
Name: SK_ID_BUREAU, Length: 305811, dtype: int64

In [23]:
print(bureau_df.columns.values)

['SK_ID_CURR' 'SK_ID_BUREAU' 'CREDIT_ACTIVE' 'CREDIT_CURRENCY'
 'DAYS_CREDIT' 'CREDIT_DAY_OVERDUE' 'DAYS_CREDIT_ENDDATE'
 'DAYS_ENDDATE_FACT' 'AMT_CREDIT_MAX_OVERDUE' 'CNT_CREDIT_PROLONG'
 'AMT_CREDIT_SUM' 'AMT_CREDIT_SUM_DEBT' 'AMT_CREDIT_SUM_LIMIT'
 'AMT_CREDIT_SUM_OVERDUE' 'CREDIT_TYPE' 'DAYS_CREDIT_UPDATE' 'AMT_ANNUITY']


One client open different credit card reporting to credit bureau

In [24]:
agg_df = pd.DataFrame(bureau_df.groupby('SK_ID_CURR')['SK_ID_BUREAU'].size()).reset_index()
agg_df.columns = ['SK_ID_CURR','BU_count']
agg_df.sort_values('BU_count',inplace=True,ascending=False)
agg_df.head()

SK_ID_CURR  BU_count
17942       120860       116
59911       169704        94
187259      318065        78
130327      251643        61
279405      425396        60

In [25]:
bureau_df.loc[bureau_df.SK_ID_CURR==120860]

SK_ID_CURR  SK_ID_BUREAU CREDIT_ACTIVE CREDIT_CURRENCY  DAYS_CREDIT  \
115133      120860       6800812        Closed      currency 1         -265   
182139      120860       6800826        Closed      currency 1         -178   
182140      120860       6800828        Closed      currency 1         -120   
182141      120860       6800830        Closed      currency 1         -148   
182142      120860       6800836        Closed      currency 1          -22   
...            ...           ...           ...             ...          ...   
880937      120860       6800880        Closed      currency 1        -1338   
880938      120860       6800881        Closed      currency 1        -1564   
880939      120860       6800882        Closed      currency 1        -2127   
880940      120860       6800883        Closed      currency 1        -1931   
880941      120860       6800884        Closed      currency 1        -2771   

        CREDIT_DAY_OVERDUE  DAYS_CREDIT_ENDDATE  DAYS_ENDDATE_FACT  \
115133                   0               -260.0             -260.0   
182139                   0               -158.0             -158.0   
182140                   0                  NaN              -76.0   
182141                   0                  NaN              -76.0   
182142                   0                  NaN              -11.0   
...                    ...                  ...                ...   
880937                   0                  NaN             -110.0   
880938                   0              -1197.0            -1197.0   
880939                   0              -1820.0            -2004.0   
880940                   0              -1624.0            -1624.0   
880941                   0              -2467.0            -2467.0   

        AMT_CREDIT_MAX_OVERDUE  CNT_CREDIT_PROLONG  AMT_CREDIT_SUM  \
115133                     NaN                   0         67500.0   
182139                     NaN                   0         67500.0   
182140                     NaN                   0         36000.0   
182141                     NaN                   0         36000.0   
182142                     NaN                   0         36000.0   
...                        ...                 ...             ...   
880937                     0.0                   0        405000.0   
880938                     0.0                   0         67500.0   
880939                     0.0                   0         39317.4   
880940                     0.0                   0         67991.4   
880941                     0.0                   0         70870.5   

        AMT_CREDIT_SUM_DEBT  AMT_CREDIT_SUM_LIMIT  AMT_CREDIT_SUM_OVERDUE  \
115133                  0.0                   NaN                     0.0   
182139                  0.0                   NaN                     0.0   
182140                  0.0                   NaN                     0.0   
182141                  0.0                   NaN                     0.0   
182142                  0.0                   NaN                     0.0   
...                     ...                   ...                     ...   
880937                  0.0                   0.0                     0.0   
880938                  NaN                   NaN                     0.0   
880939                  0.0                   0.0                     0.0   
880940                  0.0                   0.0                     0.0   
880941                  0.0                   0.0                     0.0   

            CREDIT_TYPE  DAYS_CREDIT_UPDATE  AMT_ANNUITY  
115133        Microloan                -248          NaN  
182139        Microloan                -139          NaN  
182140        Microloan                 -76          0.0  
182141        Microloan                 -76          0.0  
182142        Microloan                 -11          NaN  
...                 ...                 ...          ...  
880937      Credit card                

1. lets look at CREDIT_DAY_OVERDUE and AMT_CREDIT_MAX_OVERDUE

In [26]:
bureau_df['CREDIT_DAY_OVERDUE']

0          0
1          0
2          0
3          0
4          0
          ..
1716423    0
1716424    0
1716425    0
1716426    0
1716427    0
Name: CREDIT_DAY_OVERDUE, Length: 1716428, dtype: int64

In [27]:
bureau_temp_df = bureau_df.loc[bureau_df.CREDIT_DAY_OVERDUE > 40]
bureau_temp_df.sort_values('AMT_CREDIT_MAX_OVERDUE',inplace=True,ascending=False)
bureau_temp_df[['SK_ID_CURR','SK_ID_BUREAU','CREDIT_DAY_OVERDUE','AMT_CREDIT_MAX_OVERDUE']]

/Users/katy/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


SK_ID_CURR  SK_ID_BUREAU  CREDIT_DAY_OVERDUE  AMT_CREDIT_MAX_OVERDUE
1573873      266765       5758166                 275             1224474.885
856019       142384       6036734                 232              959386.500
1334087      427996       6600530                 178              888840.000
374695       142384       6036732                 237              883512.000
804300       340354       6512568                1742              573748.740
...             ...           ...                 ...                     ...
1701796      314436       6839208                 174                     NaN
1702134      153931       6839653                  60                     NaN
1702959      192270       6840618                  61                     NaN
1703093      372475       6840793                  41                     NaN
1713547      177331       5085700                  60                     NaN

[1914 rows x 4 columns]

supposed credit duration:

In [28]:
# supposed credit duration:
bureau_df['SUPPOSED_DURATION_CREDIT'] = bureau_df['DAYS_CREDIT_ENDDATE'] - bureau_df['DAYS_CREDIT']
bureau_df['SUPPOSED_DURATION_CREDIT']

0           344.0
1          1283.0
2           731.0
3             NaN
4          1826.0
            ...  
1716423      14.0
1716424     215.0
1716425     181.0
1716426     365.0
1716427       NaN
Name: SUPPOSED_DURATION_CREDIT, Length: 1716428, dtype: float64

actual credit duration:

In [29]:
bureau_df['ACTUAL_DURATION_CREDIT'] = bureau_df['DAYS_ENDDATE_FACT'] - bureau_df['DAYS_CREDIT']
bureau_df['ACTUAL_DURATION_CREDIT']

0          344.0
1            NaN
2            NaN
3            NaN
4            NaN
           ...  
1716423      NaN
1716424    155.0
1716425    839.0
1716426    365.0
1716427     76.0
Name: ACTUAL_DURATION_CREDIT, Length: 1716428, dtype: float64

diff between credit duration and actuall duration:

In [30]:
bureau_df['DIFF_DURATION_CREDIT'] = bureau_df['ACTUAL_DURATION_CREDIT'] - bureau_df['SUPPOSED_DURATION_CREDIT']
bureau_df['DIFF_DURATION_CREDIT'] 

0            0.0
1            NaN
2            NaN
3            NaN
4            NaN
           ...  
1716423      NaN
1716424    -60.0
1716425    658.0
1716426      0.0
1716427      NaN
Name: DIFF_DURATION_CREDIT, Length: 1716428, dtype: float64

### Assignment 3: team up and create features for credit_balance

# bureau_balance

Status of Credit Bureau loan during the month
C means closed, 
X means status unknown, 
0 means no DPD, 
1 means maximal did during month between 1-30, 
2 means DPD 31-60,… 5 means DPD 120+ or sold or written off 

In [31]:
bureau_balance_df.head()

SK_ID_BUREAU  MONTHS_BALANCE STATUS
0       5715448               0      C
1       5715448              -1      C
2       5715448              -2      C
3       5715448              -3      C
4       5715448              -4      C

lets create a new feature called BB_NO_DPD, if 'STATUS' == 0 

In [32]:
bureau_balance_df['BB_NO_DPD'] = bureau_balance_df['STATUS'].apply(lambda x: True if x==0 else False)
bureau_balance_df[['SK_ID_BUREAU', 'BB_NO_DPD']]

SK_ID_BUREAU  BB_NO_DPD
0              5715448      False
1              5715448      False
2              5715448      False
3              5715448      False
4              5715448      False
...                ...        ...
27299920       5041336      False
27299921       5041336      False
27299922       5041336      False
27299923       5041336      False
27299924       5041336      False

[27299925 rows x 2 columns]

# credit_balance

In [33]:
credit_balance_df.head()

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  AMT_BALANCE  \
0     2562384      378907              -6       56.970   
1     2582071      363914              -1    63975.555   
2     1740877      371185              -7    31815.225   
3     1389973      337855              -4   236572.110   
4     1891521      126868              -1   453919.455   

   AMT_CREDIT_LIMIT_ACTUAL  AMT_DRAWINGS_ATM_CURRENT  AMT_DRAWINGS_CURRENT  \
0                   135000                       0.0                 877.5   
1                    45000                    2250.0                2250.0   
2                   450000                       0.0                   0.0   
3                   225000                    2250.0                2250.0   
4                   450000                       0.0               11547.0   

   AMT_DRAWINGS_OTHER_CURRENT  AMT_DRAWINGS_POS_CURRENT  \
0                         0.0                     877.5   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                   11547.0   

   AMT_INST_MIN_REGULARITY  ...  AMT_RECIVABLE  AMT_TOTAL_RECEIVABLE  \
0                 1700.325  ...          0.000                 0.000   
1                 2250.000  ...      64875.555             64875.555   
2                 2250.000  ...      31460.085             31460.085   
3                11795.760  ...     233048.970            233048.970   
4                22924.890  ...     453919.455            453919.455   

   CNT_DRAWINGS_ATM_CURRENT  CNT_DRAWINGS_CURRENT  CNT_DRAWINGS_OTHER_CURRENT  \
0                       0.0                     1                         0.0   
1                       1.0                     1                         0.0   
2                       0.0                     0                         0.0   
3                       1.0                     1                         0.0   
4                       0.0                     1                         0.0   

   CNT_DRAWINGS_POS_CURRENT  CNT_INSTALMENT_MATURE_CUM  NAME_CONTRACT_STATUS  \
0                       1.0                       35.0                Active   
1                       0.0                       69.0                Active   
2                       0.0                       30.0                Active   
3                       0.0                       10.0                Active   
4                       1.0                      101.0                Active   

   SK_DPD  SK_DPD_DEF  
0       0           0  
1       0           0  
2       0           0  
3       0           0  
4       0           0  

[5 rows x 23 columns]

In [34]:
credit_balance_df.columns

Index(['SK_ID_PREV', 'SK_ID_CURR', 'MONTHS_BALANCE', 'AMT_BALANCE',
       'AMT_CREDIT_LIMIT_ACTUAL', 'AMT_DRAWINGS_ATM_CURRENT',
       'AMT_DRAWINGS_CURRENT', 'AMT_DRAWINGS_OTHER_CURRENT',
       'AMT_DRAWINGS_POS_CURRENT', 'AMT_INST_MIN_REGULARITY',
       'AMT_PAYMENT_CURRENT', 'AMT_PAYMENT_TOTAL_CURRENT',
       'AMT_RECEIVABLE_PRINCIPAL', 'AMT_RECIVABLE', 'AMT_TOTAL_RECEIVABLE',
       'CNT_DRAWINGS_ATM_CURRENT', 'CNT_DRAWINGS_CURRENT',
       'CNT_DRAWINGS_OTHER_CURRENT', 'CNT_DRAWINGS_POS_CURRENT',
       'CNT_INSTALMENT_MATURE_CUM', 'NAME_CONTRACT_STATUS', 'SK_DPD',
       'SK_DPD_DEF'],
      dtype='object')

Note: please pay more attention to col: "MONTHS_BALANCE" per user, because it indicates the date of the credit and balance away from current application. I suggest that focus on one user and apply your idea to all users after. This is time series data

In [35]:
# let us take ID 378907 as example
temp_df = credit_balance_df.loc[credit_balance_df.SK_ID_CURR==378907]
temp_df.sort_values('MONTHS_BALANCE',inplace=True,ascending=False)

/Users/katy/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [36]:
temp_df.head()

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  AMT_BALANCE  \
330701      2562384      378907              -2       28.575   
2070436     2562384      378907              -3     1897.875   
1548947     2562384      378907              -4     4036.860   
1821907     2562384      378907              -5     5228.280   
0           2562384      378907              -6       56.970   

         AMT_CREDIT_LIMIT_ACTUAL  AMT_DRAWINGS_ATM_CURRENT  \
330701                      9000                       0.0   
2070436                     9000                       0.0   
1548947                     9000                       0.0   
1821907                     9000                       0.0   
0                         135000                       0.0   

         AMT_DRAWINGS_CURRENT  AMT_DRAWINGS_OTHER_CURRENT  \
330701                    0.0                         0.0   
2070436                   0.0                         0.0   
1548947                 900.0                         0.0   
1821907                5229.0                         0.0   
0                       877.5                         0.0   

         AMT_DRAWINGS_POS_CURRENT  AMT_INST_MIN_REGULARITY  ...  \
330701                        0.0                 1916.910  ...   
2070436                       0.0                 2250.000  ...   
1548947                     900.0                 2250.000  ...   
1821907                    5229.0                  840.780  ...   
0                           877.5                 1700.325  ...   

         AMT_RECIVABLE  AMT_TOTAL_RECEIVABLE  CNT_DRAWINGS_ATM_CURRENT  \
330701          28.575                28.575                       0.0   
2070436       1897.875              1897.875                       0.0   
1548947       4036.860              4036.860                       0.0   
1821907       5228.280              5228.280                       0.0   
0                0.000                 0.000                       0.0   

         CNT_DRAWINGS_CURRENT  CNT_DRAWINGS_OTHER_CURRENT  \
330701                      0                         0.0   
2070436                     0                         0.0   
1548947                     1                         0.0   
1821907                     2                         0.0   
0                           1                         0.0   

         CNT_DRAWINGS_POS_CURRENT  CNT_INSTALMENT_MATURE_CUM  \
330701                        0.0                       39.0   
2070436                       0.0                       38.0   
1548947                       1.0                       37.0   
1821907                       2.0                       36.0   
0                             1.0                       35.0   

         NAME_CONTRACT_STATUS  SK_DPD  SK_DPD_DEF  
330701                 Active       0           0  
2070436                Active       0           0  
1548947                Active       0           0  
1821907                Active       0           0  
0                      Active       0           0  

[5 rows x 23 columns]

### Assignment 4: team up and create features for installment history

# installment.csv

In [37]:
installments_df.loc[installments_df.AMT_INSTALMENT!=installments_df.AMT_PAYMENT]
installments_df[['SK_ID_PREV', 'AMT_PAYMENT']]

SK_ID_PREV  AMT_PAYMENT
0            1054186     6948.360
1            1330831     1716.525
2            2085231    25425.000
3            2452527    24350.130
4            2714724     2160.585
...              ...          ...
13605396     2186857          NaN
13605397     1310347          NaN
13605398     1308766          NaN
13605399     1062206          NaN
13605400     2448869          NaN

[13605401 rows x 2 columns]

In [38]:
installments_df[['SK_ID_PREV', 'AMT_INSTALMENT']]

SK_ID_PREV  AMT_INSTALMENT
0            1054186        6948.360
1            1330831        1716.525
2            2085231       25425.000
3            2452527       24350.130
4            2714724        2165.040
...              ...             ...
13605396     2186857          67.500
13605397     1310347          67.500
13605398     1308766       43737.435
13605399     1062206          67.500
13605400     2448869       11504.250

[13605401 rows x 2 columns]

lets see if the difference between AMT_INSTALMENT and AMT_PAYMENT, give us anything intresting 

In [39]:
installments_df['AMT_INSTALMENT_DIFF'] = installments_df[ 'AMT_INSTALMENT'] - installments_df[ 'AMT_PAYMENT']

In [40]:
installments_df[['SK_ID_PREV', 'AMT_INSTALMENT_DIFF']]

SK_ID_PREV  AMT_INSTALMENT_DIFF
0            1054186                0.000
1            1330831                0.000
2            2085231                0.000
3            2452527                0.000
4            2714724                4.455
...              ...                  ...
13605396     2186857                  NaN
13605397     1310347                  NaN
13605398     1308766                  NaN
13605399     1062206                  NaN
13605400     2448869                  NaN

[13605401 rows x 2 columns]

Lets see the difference between: 
1. DAYS_ENTRY_PAYMENT - When was the installments of previous credit paid actually (relative to application date of current loan)
2. DAYS_INSTALMENT- When the installment of previous credit was supposed to be paid (relative to application date of current loan)

In [41]:
installments_df[ 'DAYS_INSTALMENT']

0          -1180.0
1          -2156.0
2            -63.0
3          -2418.0
4          -1383.0
             ...  
13605396   -1624.0
13605397   -1539.0
13605398      -7.0
13605399   -1986.0
13605400     -27.0
Name: DAYS_INSTALMENT, Length: 13605401, dtype: float64

In [42]:
installments_df['DAYS_INSTALMENT_DIFF'] = (installments_df[ 'DAYS_ENTRY_PAYMENT'] - installments_df[ 'DAYS_INSTALMENT']).clip(lower=0)
installments_df['DAYS_INSTALMENT_DIFF']

0            0.0
1            0.0
2            0.0
3            0.0
4           17.0
            ... 
13605396     NaN
13605397     NaN
13605398     NaN
13605399     NaN
13605400     NaN
Name: DAYS_INSTALMENT_DIFF, Length: 13605401, dtype: float64

In [43]:
installments_df[['SK_ID_CURR', 'SK_ID_PREV', 'DAYS_INSTALMENT_DIFF']]

SK_ID_CURR  SK_ID_PREV  DAYS_INSTALMENT_DIFF
0             161674     1054186                   0.0
1             151639     1330831                   0.0
2             193053     2085231                   0.0
3             199697     2452527                   0.0
4             167756     2714724                  17.0
...              ...         ...                   ...
13605396      428057     2186857                   NaN
13605397      414406     1310347                   NaN
13605398      402199     1308766                   NaN
13605399      409297     1062206                   NaN
13605400      434321     2448869                   NaN

[13605401 rows x 3 columns]

DAYS_INSTALMENT, which indidates the days of paying installent, you need to sort them by each client

In [44]:
temp_installments_df = installments_df.loc[installments_df.SK_ID_CURR==378907]
temp_installments_df.sort_values('DAYS_INSTALMENT',inplace=True,ascending=False)

/Users/katy/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


let see it for SK_ID_CURR==378907

In [45]:
temp_installments_df.head()

SK_ID_PREV  SK_ID_CURR  NUM_INSTALMENT_VERSION  \
10225387     2562384      378907                     0.0   
10288776     2562384      378907                     0.0   
9059308      2562384      378907                     0.0   
8937336      2562384      378907                     0.0   
11352964     2562384      378907                     0.0   

          NUM_INSTALMENT_NUMBER  DAYS_INSTALMENT  DAYS_ENTRY_PAYMENT  \
10225387                     77            -32.0               -32.0   
10288776                     76            -63.0               -61.0   
9059308                      75            -91.0               -91.0   
8937336                      74           -122.0              -117.0   
11352964                     72           -153.0              -153.0   

          AMT_INSTALMENT  AMT_PAYMENT  AMT_INSTALMENT_DIFF  \
10225387          28.575       28.575                  0.0   
10288776        1916.910     1916.910                  0.0   
9059308         2250.000     2250.000                  0.0   
8937336         2250.000     2250.000                  0.0   
11352964         840.780      840.780                  0.0   

          DAYS_INSTALMENT_DIFF  
10225387                   0.0  
10288776                   2.0  
9059308                    0.0  
8937336                    5.0  
11352964                   0.0

### Assignment 5: complete feature engineering by sorting by "MONTHS_BALANCE"


# pos_cash.csv

In [46]:
pos_cash_df.head()

SK_ID_PREV  SK_ID_CURR  MONTHS_BALANCE  CNT_INSTALMENT  \
0     1803195      182943             -31            48.0   
1     1715348      367990             -33            36.0   
2     1784872      397406             -32            12.0   
3     1903291      269225             -35            48.0   
4     2341044      334279             -35            36.0   

   CNT_INSTALMENT_FUTURE NAME_CONTRACT_STATUS  SK_DPD  SK_DPD_DEF  
0                   45.0               Active       0           0  
1                   35.0               Active       0           0  
2                    9.0               Active       0           0  
3                   42.0               Active       0           0  
4                   35.0               Active       0           0

1. SK_DPD refers to the dpd (days past due) for any amounts(even if it is small); 
2. SK_DPD_DEF refers to the dpd (days past due) for those relatively "significant" amounts. Which saying, we should take SK_DPD_DEF as the ideal column for evaluating the customer's dpd behaviors.
3. SK_DPD is often bigger than SK_DPD_DEF
******
4. New Feature: Created a new feature called "SK_DPD_DEF_risk" has values 0,1, One(1) for high risk, Zero(0) for low risk applicant. (High Risk - applicants with more than 30 days -ast due. ) 

In [47]:
pos_cash_df['SK_DPD_DEF_risk'] = (pos_cash_df['SK_DPD_DEF'] > 30).astype(int)
pos_cash_df[['SK_ID_PREV','SK_DPD_DEF_risk']]

SK_ID_PREV  SK_DPD_DEF_risk
0            1803195                0
1            1715348                0
2            1784872                0
3            1903291                0
4            2341044                0
...              ...              ...
10001353     2448283                0
10001354     1717234                0
10001355     1283126                0
10001356     1082516                0
10001357     1259607                0

[10001358 rows x 2 columns]

1. We know that SK_DPD is often bigger than SK_DPD_DEF, lets create New Freature call "SK_DPD_diff"

In [48]:
pos_cash_df['SK_DPD_diff'] = pos_cash_df['SK_DPD'] - pos_cash_df['SK_DPD_DEF']
pos_cash_df[['SK_ID_PREV','SK_DPD_diff']]

SK_ID_PREV  SK_DPD_diff
0            1803195            0
1            1715348            0
2            1784872            0
3            1903291            0
4            2341044            0
...              ...          ...
10001353     2448283          843
10001354     1717234          602
10001355     1283126          609
10001356     1082516          614
10001357     1259607            0

[10001358 rows x 2 columns]

In [49]:
a = sorted(pos_cash_df['MONTHS_BALANCE'])
print(a)
#pos_cash_df[['SK_ID_PREV','MONTHS_BALANCE']]

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



### Assignment 6: complete feature engineering for this table
#### prev_df.csv
This file shows the previous activity for clients in the bank(not in other banks), you can independently process it and merge into the main table as new features

In [50]:
prev_df.head()

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START  ... NAME_SELLER_INDUSTRY  CNT_PAYMENT  \
0                       15  ...         Connectivity         12.0   
1                       11  ...                  XNA         36.0   
2                       11  ...                  XNA         12.0   
3                        7  ...                  XNA         12.0   
4                        9  ...                  XNA         24.0   

   NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0            middle  POS mobile with interest            365243.0   
1        low_action          Cash X-Sell: low            365243.0   
2              high         Cash X-Sell: high            365243.0   
3            middle       Cash X-Sell: middle            365243.0   
4              high         Cash Street: high                 NaN   

  DAYS_FIRST_DUE DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
0          -42.0                     300.0          -42.0            -37.0   
1         -134.0                     916.0       365243.0         365243.0   
2         -271.0                      59.0       365243.0         365243.0   
3         -482.0                    -152.0         -182.0           -177.0   
4            NaN                       NaN            NaN              NaN   

  NFLAG_INSURED_ON_APPROVAL  
0                       0.0  
1                       1.0  
2                       1.0  
3                       1.0  
4                       NaN  

[5 rows x 37 columns]

here are some features I made to help you


1. AMT_CREDIT - Final credit amount on the previous application. This differs from AMT_APPLICATION in a way that the AMT_APPLICATION is the amount for which the client initially applied for, but during our approval process he could have received different amount - AMT_CREDIT

2. AMT_APPLICATION	For how much credit did client ask on the previous application
3. Creating new feature called: 'RATE_AMT_CREDIT' = how much client asked / final amount recieved, if the number bigger than 1, meaning he is a preferable client. 


In [51]:
prev_df['RATE_AMT_CREDIT'] = prev_df['AMT_APPLICATION']/prev_df['AMT_CREDIT']
prev_df[['SK_ID_PREV', 'RATE_AMT_CREDIT']]

SK_ID_PREV  RATE_AMT_CREDIT
0           2030495         1.000000
1           2802425         0.893815
2           2523466         0.824511
3           2819243         0.955840
4           1784265         0.835282
...             ...              ...
1670209     2300464         0.858367
1670210     2357031         1.364877
1670211     2659632         1.026467
1670212     2785582         0.938086
1670213     2418762         1.000000

[1670214 rows x 2 columns]

1. AMT_CREDIT - Final credit amount on the previous application. This differs from AMT_APPLICATION in a way that the AMT_APPLICATION is the amount for which the client initially applied for, but during our approval process he could have received different amount - AMT_CREDIT
2. AMT_ANNUITY- Annuity (a fixed sum of money paid to someone each year) of previous application
3. Creating new feature called: 'RATE_ANN_CREDIT' = Annuity of previous application / final amount recieved

In [52]:
prev_df['RATE_ANN_CREDIT'] = prev_df['AMT_ANNUITY']/prev_df['AMT_CREDIT']
prev_df[['SK_ID_PREV', 'RATE_ANN_CREDIT']]

SK_ID_PREV  RATE_ANN_CREDIT
0           2030495         0.100929
1           2802425         0.037060
2           2523466         0.110380
3           2819243         0.099920
4           1784265         0.079010
...             ...              ...
1670209     2300464         0.047220
1670210     2357031         0.103000
1670211     2659632         0.112373
1670212     2785582         0.098090
1670213     2418762         0.045642

[1670214 rows x 2 columns]

1. 'AMT_DOWN_PAYMENT' - Down payment on the previous application
2. AMT_CREDIT - Final credit amount on the previous application. This differs from AMT_APPLICATION in a way that the AMT_APPLICATION is the amount for which the client initially applied for, but during our approval process he could have received different amount - AMT_CREDIT
3. Creating new feature called: 'RATE_DOWNPAY_CREDIT' = Down payment on the previous application / Final credit amount on the previous application, 'RATE_DOWNPAY_CREDIT' would be one (1) if it's less or equal to 0, otherwise will be one (0)


In [53]:
prev_df['RATE_DOWNPAY_CREDIT'] = prev_df['AMT_DOWN_PAYMENT']/prev_df['AMT_CREDIT']
prev_df[['SK_ID_PREV', 'RATE_DOWNPAY_CREDIT']]

SK_ID_PREV  RATE_DOWNPAY_CREDIT
0           2030495             0.000000
1           2802425                  NaN
2           2523466                  NaN
3           2819243                  NaN
4           1784265                  NaN
...             ...                  ...
1670209     2300464             0.000000
1670210     2357031             0.454959
1670211     2659632             0.102664
1670212     2785582                  NaN
1670213     2418762                  NaN

[1670214 rows x 2 columns]

In [54]:
prev_df['RATE_DOWNPAY_CREDIT'] = prev_df['RATE_DOWNPAY_CREDIT'].apply(lambda x: 1 if x < 0 or x ==0  else 0)
prev_df[['SK_ID_PREV', 'RATE_DOWNPAY_CREDIT']]

SK_ID_PREV  RATE_DOWNPAY_CREDIT
0           2030495                    1
1           2802425                    0
2           2523466                    0
3           2819243                    0
4           1784265                    0
...             ...                  ...
1670209     2300464                    1
1670210     2357031                    0
1670211     2659632                    0
1670212     2785582                    0
1670213     2418762                    0

[1670214 rows x 2 columns]

1. AMT_ANNUITY - Annuity (a fixed sum of money paid to someone each year) of previous application
2. AMT_APPLICATION - is the amount for which the client initially applied for, but during our approval process he could have received different amount - AMT_CREDIT
3. Creating new feature called: 'RATE_ANN_APPT' = Annuity of previous application / amount client initially applied for

In [55]:
prev_df['RATE_ANN_APP'] = prev_df['AMT_ANNUITY']/prev_df['AMT_APPLICATION']
prev_df[['SK_ID_PREV', 'RATE_ANN_APP']]

SK_ID_PREV  RATE_ANN_APP
0           2030495      0.100929
1           2802425      0.041463
2           2523466      0.133873
3           2819243      0.104536
4           1784265      0.094591
...             ...           ...
1670209     2300464      0.055011
1670210     2357031      0.075465
1670211     2659632      0.109475
1670212     2785582      0.104564
1670213     2418762      0.045642

[1670214 rows x 2 columns]

In [56]:
# In real scenario it would be useful to encode 'CODE_REJECT_REASON', but I have no way of encoding and understanding XAP or HC 
prev_df[['SK_ID_PREV', 'CODE_REJECT_REASON']]

SK_ID_PREV CODE_REJECT_REASON
0           2030495                XAP
1           2802425                XAP
2           2523466                XAP
3           2819243                XAP
4           1784265                 HC
...             ...                ...
1670209     2300464                XAP
1670210     2357031                XAP
1670211     2659632                XAP
1670212     2785582                XAP
1670213     2418762                XAP

[1670214 rows x 2 columns]

Was the previous application for CASH, POS, CAR, …
This might help to find out which have own a car

In [57]:
prev_df[['SK_ID_PREV', 'NAME_PORTFOLIO']]

SK_ID_PREV NAME_PORTFOLIO
0           2030495            POS
1           2802425           Cash
2           2523466           Cash
3           2819243           Cash
4           1784265           Cash
...             ...            ...
1670209     2300464            POS
1670210     2357031            POS
1670211     2659632            POS
1670212     2785582           Cash
1670213     2418762           Cash

[1670214 rows x 2 columns]

Through which channel we acquired the client on the previous application
This might be helpful for marketing

In [58]:
prev_df[['SK_ID_PREV', 'CHANNEL_TYPE']]

SK_ID_PREV             CHANNEL_TYPE
0           2030495             Country-wide
1           2802425           Contact center
2           2523466  Credit and cash offices
3           2819243  Credit and cash offices
4           1784265  Credit and cash offices
...             ...                      ...
1670209     2300464                    Stone
1670210     2357031                    Stone
1670211     2659632             Country-wide
1670212     2785582          AP+ (Cash loan)
1670213     2418762          AP+ (Cash loan)

[1670214 rows x 2 columns]

NAME_TYPE_SUITE - Who accompanied client when applying for the previous application, we can create a new feature to find who needed a co-signer, We assume that those with accompany needed a Co-Signer. 

In [59]:
prev_df[['SK_ID_PREV', 'NAME_TYPE_SUITE']]

SK_ID_PREV  NAME_TYPE_SUITE
0           2030495              NaN
1           2802425    Unaccompanied
2           2523466  Spouse, partner
3           2819243              NaN
4           1784265              NaN
...             ...              ...
1670209     2300464              NaN
1670210     2357031    Unaccompanied
1670211     2659632  Spouse, partner
1670212     2785582           Family
1670213     2418762           Family

[1670214 rows x 2 columns]

Create Co_SIGNER flag - True for having co-signer and false for not having
Note: this an assumption, in real case we do more investigation 

In [60]:
prev_df['CO_SIGNER'] = prev_df['NAME_TYPE_SUITE'].apply(lambda x: False if x is np.nan else True)
prev_df[['SK_ID_PREV', 'CO_SIGNER']]


SK_ID_PREV  CO_SIGNER
0           2030495      False
1           2802425       True
2           2523466       True
3           2819243      False
4           1784265      False
...             ...        ...
1670209     2300464      False
1670210     2357031       True
1670211     2659632       True
1670212     2785582       True
1670213     2418762       True

[1670214 rows x 2 columns]